apps/src/pythonlab/pyodideWorkerManager.js

```diff
@@ -10,59 +10,77 @@ import {setAndSaveProjectSource} from '@cdo/apps/lab2/redux/lab2ProjectRedux';
 import MetricsReporter from '@cdo/apps/lib/metrics/MetricsReporter';
 import {getStore} from '@cdo/apps/redux';
 
+import {setLoadingCodeEnvironment} from '../lab2/redux/systemRedux';
+
 import {parseErrorMessage} from './pythonHelpers/messageHelpers';
 import {MATPLOTLIB_IMG_TAG} from './pythonHelpers/patches';
 
-// This syntax doesn't work with typescript, so this file is in js.
-const pyodideWorker = new Worker(
-  new URL('./pyodideWebWorker.js', import.meta.url)
-);
+let callbacks = {};
+
+const setUpPyodideWorker = () => {
+  // This syntax doesn't work with typescript, so this file is in js.
+  const worker = new Worker(new URL('./pyodideWebWorker.js', import.meta.url));
 
-const callbacks = {};
+  callbacks = {};
 
-pyodideWorker.onmessage = event => {
-  const {type, id, message} = event.data;
-  if (type === 'sysout' || type === 'syserr') {
-    if (message.startsWith(MATPLOTLIB_IMG_TAG)) {
-      // This is a matplotlib image, so we need to append it to the output
-      const image = message.slice(MATPLOTLIB_IMG_TAG.length + 1);
-      getStore().dispatch(appendOutputImage(image));
+  worker.onmessage = event => {
+    const {type, id, message} = event.data;
+    if (type === 'sysout' || type === 'syserr') {
+      if (message.startsWith(MATPLOTLIB_IMG_TAG)) {
+        // This is a matplotlib image, so we need to append it to the output
+        const image = message.slice(MATPLOTLIB_IMG_TAG.length + 1);
+        getStore().dispatch(appendOutputImage(image));
+        return;
+      }
+      getStore().dispatch(appendSystemOutMessage(message));
+      return;
+    } else if (type === 'run_complete') {
+      getStore().dispatch(appendSystemMessage('Program completed.'));
+    } else if (type === 'updated_source') {
+      getStore().dispatch(setAndSaveProjectSource({source: message}));
+      return;
+    } else if (type === 'error') {
+      getStore().dispatch(appendErrorMessage(parseErrorMessage(message)));
+      return;
+    } else if (type === 'internal_error') {
+      MetricsReporter.logError({
+        type: 'PythonLabInternalError',
+        message,
+      });
+      return;
+    } else if (type === 'system_error') {
+      getStore().dispatch(appendSystemError(message));
+      MetricsReporter.logError({
+        type: 'PythonLabSystemCodeError',
+        message,
+      });
+      return;
+    } else if (type === 'loading_pyodide') {
+      getStore().dispatch(
+        appendSystemMessage('Loading your python environment...')
+      );
+      getStore().dispatch(setLoadingCodeEnvironment(true));
+      return;
+    } else if (type === 'loaded_pyodide') {
+      getStore().dispatch(appendSystemMessage('Your environment is ready!'));
+      getStore().dispatch(setLoadingCodeEnvironment(false));
+      return;
+    } else {
```

Comment:
I only suggest this when an if/then chain is getting really long, but at a certain point I think its a candidate for a `switch` statement. This clarifies at first glance that only the `type` will be used for varying conditions, and thus clarifies the relative simplicity of the long stanza.

Also, switch statements have a way of avoiding future maintainers years down the road from ADDING complexity by if/then-ing on complex conditions and stretching the simplicity of the message forking code.

The only except I see here is the merging of sysout and syserr, I would argue these would be well served by extracting the contents of the block and having two separate one-liner sysout and syserr case statements.

Reply:
Not necessarily a candidate idea for this PR or particularly important, more a general thought.

Reply:
I'll update this as a follow up!

Reply:
https://codedotorg.atlassian.net/browse/CT-730